python-api/livy/job_handle.py

```diff
@@ -0,0 +1,257 @@
+                suffix_url = "/" + str(self._session_id) + "/jobs/" + str(
+                    self._job_id)
+                headers = {'X-Requested-By': 'livy'}
+                job_status = self._conn.send_request('GET', suffix_url,
+                                                     headers=headers).json()
+                job_state = job_status['state']
+                job_result = None
+                has_finished = False
+                job_error = None
+                if job_state == 'SUCCEEDED':
+                    job_result = job_status['result']
+                    has_finished = True
+                elif job_state == 'FAILED':
+                    job_error = job_status['error']
+                    has_finished = True
+                elif job_state == 'CANCELLED':
+                    repeated_timer.stop()
+                else:
+                    pass
+                if has_finished:
+                    if job_result is not None:
+                        b64_decoded = base64.b64decode(job_result)
+                        b64_decoded_decoded = base64.b64decode(b64_decoded)
+                        deserialized_object = cloudpickle.loads(
+                            b64_decoded_decoded)
+                        super(JobHandle, self).set_result(deserialized_object)
+                    if job_error is not None:
+                        super(JobHandle, self).set_exception_info(Exception(
+                            job_error), None)
+                    repeated_timer.stop()
+                else:
+                    self._update_state(job_state)
+            except Exception as err:
+                repeated_timer.stop()
+                traceback.print_exc()
+                super(JobHandle, self).set_exception_info(
+                    err, traceback.format_exc())
+
+        repeated_timer = self._RepeatedTimer(
+            self._JOB_INITIAL_POLL_INTERVAL, do_poll_result)
+        repeated_timer.start()
+
+    def set_running_or_notify_cancel(self):
+        raise NotImplementedError("This operation is not supported.")
+
+    def set_result(self, result):
+        raise NotImplementedError("This operation is not supported.")
+
+    def set_exception_info(self, exception, traceback):
+        raise NotImplementedError("This operation is not supported.")
+
+    def set_exception(self, exception):
+        raise NotImplementedError("This operation is not supported.")
+
+    class _RepeatedTimer(object):
+        def __init__(self, interval, polling_job):
+            self._timer = None
+            self.polling_job = polling_job
+            self.interval = interval
+            self.is_running = False
+            self.stop_called = False
+
+        def _run(self):
+            self.is_running = False
+            self.polling_job()
+            self.start()
+
+        def start(self):
+            if not self.is_running and not self.stop_called:
+                self._timer = Timer(self.interval, self._run)
```

Comment:
Not a huge deal, but it would be better if this used the same executor you already instantiated, instead of creating a separate thread for each job.